CIP-0113? | Programmable token-like assets (#444)
cardano-foundation/CIPs: CIP-meta-assets (ERC20-like assets)/README.md

---
CIP: ?

> Title: ERC20-like assets

Reviewer: ```suggestion
Title: Programmable token-like assets
```
Of course @michele-nuzzi as author, or other community members, are welcome to provide alternative names... this is just a pilot to see how it looks everywhere.

>
> whith the goal of keeping the standard interoperable and easy to understand and implement
> in this first implementation we restricts transfers from a single account to a single account
>
> if necessary; this restriction might be dropped in a future version of the CIP 

The implementation requires

- a parametrized minting policy to validate the creation of new accounts (in this CIP also referred as `accountFactory`)
- a spending validator to manage the accounts ((in this CIP also referred as `accountManager`))

### standard data types used

We'll need to use some data types as defined in the script context;

```ts
const PTxId = pstruct({
    PTxId: { txId: bs }
});

const PTxOutRef = pstruct({
    PTxOutRef: {
        id: PTxId.type,
        index: int
    }
});

const PCredential = pstruct({
    PPubKeyCredential: { pkh: bs },
    PScriptCredential: { valHash: bs },
});

const PCurrencySymbol = palias( bs );
```
which are equivalent to the data:
```hs
-- PTxId
Constr 0 [ B txId ]

--- PTxOutRef
Constr 0 [ PTxId, I index ]

-- PPubKeyCredential
Constr 0 [ B pkh ]

-- PScriptCredential
Constr 1 [ B valHash ]

-- PCurrencySymbol
B currencySym
```

> ## `accountFactory` (minting policy)

Reviewer 2: ```suggestion
### `accountFactory` (minting policy)
```

Reviewer: @michele-nuzzi this is a straightforward header problem, and mandatory to fix, so let's commit it & move on to more significant issues 🤓

- the output going to the `accountManager` validator MUST implement the following checks (in addition to the presence of the minted asset):
    - the address' payment credential MUST have `ScriptCredential` constructor (constructor index `1`) and validator hash MUST equal the hard coded `accountManager` hash
    - the value attached to the UTxO must only have 2 entries (ADA and minted token) to prevent DoS by token spam.
    - the output datum MUST:
        - be constructed using the `InlineDatum` consturctor (consturctor index `2`)
        - have datum value with the [`Account`](#Account) structure (explained below) with the foeeowing fields:
            - amount: `0`
            - currencySym: currency symbol of the `accountFactory` minting policy
            - credentials: the first resolved input address credentials (both public key hash and validator hash supported)
            - state: no restrictions (up to the specific implementaiton).

optionally a specific implementation might require a fixed number of outputs (eg. a single output) for performance reasons.

> ## `accountManager` (spending validator)

Reviewer 2: ```suggestion
### `accountManager` (spending validator)
```

Reviewer: Likewise @michele-nuzzi with https://github.com/cardano-foundation/CIPs/pull/444#discussion_r1480346321.

---
CIP: 113
Title: Programmable token-like assets
Category: Tokens
Status: Proposed
Authors:
    - Michele Nuzzi <michele.nuzzi.204@gmail.com>
    - Harmonic Laboratories <harmoniclabs@protonmail.com>
    - Matteo Coppola <m.coppola.mazzetti@gmail.com>
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/444
Created: 2023-01-14
License: CC-BY-4.0
---

## Abstract

> This CIP proposes a standard that if adopted would allow the same level of programmability of other ecosistems at the price of the token true ownership.

Reviewer 3: > ecosistems

Minor spelling error, should be "ecosystems"

This is achieved imitating the way the account model works laveraging the UTxO structure adopted by Cardano.

## Motivation: why is this CIP necessary?

This CIP proposes a solution at the Cardano Problem Statement 3 ([CPS-0003](https://github.com/cardano-foundation/CIPs/pull/382/files?short_path=5a0dba0#diff-5a0dba075d998658d72169818d839f4c63cf105e4d6c3fe81e46b20d5fd3dc8f)).

If adopted it would allow to introduce the programmability over the transfer of tokens (meta-tokens) and their lifecycle.

The solution proposed includes (answering to the open questions of CPS-0003):

1 and 2) very much like account based models, wallets supporting this standard will require to know the address of the smart contract (validator)

> 3) the soultion can co-exsist with the exsiting native tokens

Reviewer 4: ```suggestion
3) the solution can co-exist with the existing native tokens
```

Maybe think about using a spell-checker in your workflow :)

> 4) the implementaiton is possible without hardfork since Vasil

Reviewer 4: ```suggestion
4) the implementation is possible without hardfork since Vasil
```

5) optimized implementations should not take significant computation, especially on transfers.

## Specification

> In the specifiaction we'll use the haskell data type `Data`:

Reviewer 4: ```suggestion
In the specification we'll use the haskell data type `Data`:
```

```hs
data Data
    = Constr Integer [Data]
    | Map [( Data, Data )]
    | List [Data]
    | I Integer
    | B ByteString
```

and we'll use the [plu-ts syntax for structs definition](https://pluts.harmoniclabs.tech/docs/onchain/Values/Structs/definition#pstruct) as an abstraction over the `Data` type.

The core idea of the implementation is to emulate the ERC20 standard; where tokens are entries in a map with addresses (or credentials in our case) as key and integers (the balances) as value. ([see the OpenZeppelin implementation for reference](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/9b3710465583284b8c4c5d2245749246bb2e0094/contracts/token/ERC20/ERC20.sol#L16));

Unlike the ERC20 standard; this CIP:

- allows for multiple entries with the same key (same credentials can be used for multiple accounts)
- DOESN'T include an equivalent of the `transferFrom` method; if needed it can be added by a specific implementation but it won't be considered part of the standard.

> **NOTE**
>
> the UTxO model allows for multiple transfers in one transaction
>
> this would allow for a more powerful model than the account based equivalent but implies higher execution costs
>

> > whith the goal of keeping the standard interoperable and easy to understand and implement

Reviewer 4: ```suggestion
> with the goal of keeping the standard interoperable and easy to understand and implement
```

> in this first implementation we restricts transfers from a single account to a single account
>
> if necessary; this restriction might be dropped in a future version of the CIP 

The implementation requires

> - a parametrized minting policy to validate the creation of new accounts (in this CIP also referred as `accountFactory`)

Reviewer 4: ```suggestion
- a parameterized minting policy to validate the creation of new accounts (in this CIP also referred as `accountFactory`)
```

> - a spending validator to manage the accounts ((in this CIP also referred as `accountManager`))

Reviewer 4: ```suggestion
- a spending validator to manage the accounts (in this CIP also referred as `accountManager`)
```

Reviewer: fixed by https://github.com/cardano-foundation/CIPs/pull/444/commits/5bf321b1e2d06425f4e0ec37aceedda86a165d85.

### standard data types used

We'll need to use some data types as defined in the script context;

```ts
const PTxId = pstruct({
    PTxId: { txId: bs }
});

const PTxOutRef = pstruct({
    PTxOutRef: {
        id: PTxId.type,
        index: int
    }
});

const PCredential = pstruct({
    PPubKeyCredential: { pkh: bs },
    PScriptCredential: { valHash: bs },
});

const PCurrencySymbol = palias( bs );
```
which are equivalent to the data:
```hs
-- PTxId
Constr 0 [ B txId ]

--- PTxOutRef
Constr 0 [ PTxId, I index ]

-- PPubKeyCredential
Constr 0 [ B pkh ]

-- PScriptCredential
Constr 1 [ B valHash ]

-- PCurrencySymbol
B currencySym
```

### `accountFactory` (minting policy)

> The `accountFactory` contract is responsabile of validating the creation of new accounts.

Reviewer 4: ```suggestion
The `accountFactory` contract is responsible for validating the creation of new accounts.
```

### validation logic

If the creation of the account is considered valid (the minting policy suceeds)

> it should result in a single minted asset going to an address with payment credentials equalst to an hard-coded (parametrized) `accountManager` contract, and stake credentials equals to the stake cretentials of the users (if any).

Reviewer 4: ```suggestion
it should result in a single minted asset going to an address with payment credentials equal to a hard-coded (parameterized) `accountManager` contract, and stake credentials equal to the stake credentials of the users (if any).
```

This done to facilitate the query of the user's accounts in the offchain.

The standard doesn't include a specification on the policy in case of burning but it is suggested to always allow for the assets burn.

> the data needed form the `ScriptContext` includes teh following fields:

Reviewer 4: ```suggestion
the data needed from the `ScriptContext` includes the following fields:
```

- `inputs`
- `output`
- `mint`

used as follows:

- all the transaction inputs MUST NOT include any tokens having the same currency symbol of the `accountFactory` minting policy;
optionally a specific implementation might require a fixed number of inputs (eg. a single input) for performance reasons.

> - when minting assets only a single token should be minted (indipendently from the number of inputs)

Reviewer 4: ```suggestion
- when minting assets only a single token should be minted (independently from the number of inputs)
```

- the only asset minted MUST have an unique asset name (we suggest using the output of `[(builtin sha2_256) [(builtin serialiseData) fstUtxoRef]]` where `fstUtxoRef` is the utxo reference (or `PTxOutRef`) of the very first input of the transaction)

> - the asset minted MUST be included in an output of the transaction being validated going to the address as specified above (payment credentials that are the hard-coded `accountManager` validator, and stake credtentials of the user if any).

Reviewer 4: ```suggestion
- the asset minted MUST be included in an output of the transaction being validated going to the address as specified above (payment credentials that are the hard-coded `accountManager` validator, and stake credentials of the user if any).
```

    - the output datum MUST:
        - be constructed using the `InlineDatum` consturctor (consturctor index `2`)
        - have datum value with the [`Account`](#Account) structure (explained below) with the foeeowing fields:
            - amount: `0`
            - currencySym: currency symbol of the `accountFactory` minting policy
            - credentials: the first resolved input address credentials (both public key hash and validator hash supported)
            - state: no restrictions (up to the specific implementaiton).

optionally a specific implementation might require a fixed number of outputs (eg. a single output) for performance reasons.

### `accountManager` (spending validator)

The `accountManager` contract is responsabile of managing exsiting accounts and their balances.

This is done using some standard redeemers and optionally implementation specfic ones.

### `Account`

The `Account` data type is used as the `accountManager` datum; and is defined as follows:

```ts
const Account = pstruct({
    Account: {
        amount: int,
        credentials: PCredential.type,
        currencySym: PCurrencySymbol.type,
        state: data
    }
});
```

> ### `AccounManagerRedeemer`

Reviewer 4: Should this be called `AccountManagerRedeemer`? Looks like a typo, but was used consistently below.

Author: yes it should, is a typo copied and pasted

Author: fixed

The `AccounManagerRedeemer` is used to comunicate the contract the intention with which the utxo is being spent.

It includes 4 standard redeemers; none of which is meant to manipulate the state of the spending account.

for this reason a specific implementation will likely have more than 4 possible redeemers that will not be considered standard
(eg. a wallet implementing an interface SHOULD NOT depend on the exsistence of these additional redeemers).

The minimal `AccounManagerRedeemer` is:

```ts
const AccounManagerRdmr = pstruct({
    Mint: { // or Burn if `amount` is negative
        amount: int
    },
    Transfer: {
        to: PCredential.type,
        amount: int
    },
    Receive: {},
    ForwardCompatibility: {}
});
```

The validation logic is different for each redeemer.

#### Common operations and values

Before proceeding with the redeemers validation logic here are some common operations and values between some of the redeemers.

##### `ownUtxoRef`

the `accountManager` contract is meant to be used only as spending validator.

> As succ we can extract the utxo being spent from the `ScriptPurpose` when constructed with the `Spending` contstructor and fail for the rest.

Reviewer 4: ```suggestion
As such we can extract the utxo being spent from the `ScriptPurpose` when constructed with the `Spending` constructor and fail for the rest.
```

```ts
const ownUtxoRef = plet(
    pmatch( ctx.purpose )
    .onSpending(({ utxoRef }) => utxoRef)
    ._( _ => perror( PTxOutRef.type ) )
);
```

##### `validatingInput`

is the input with `utxoRef` field equivalent to `ownUtxoRef`

##### `ownCreds`

from the `validatingInput`:

```ts
validatingInput.resolved.address.credential
```

##### `ownValue`

from the `validatingInput`:

```ts
validatingInput.resolved.value
```

##### `isOwnOutput`

> given an output; we recongize the output as "own" if the attached credantials are equivalent to `ownCreds` 

Reviewer 4: ```suggestion
given an output; we recongize the output as "own" if the attached credentials are equivalent to `ownCreds` 
```

        .and(
            out.value.some( ({ fst: policy }) => policy.eq( account.currencySym ) )
        ) 
    )
);
```

##### `isOwnInput` 

an input is "own" if the resolved field is "own";

```ts
const isOwnInput = plet(
    pfn([ PTxInInfo.type ], bool )
    ( input => isOwnOutput.$( input.resolved ) )
);
```

##### `isOwnCurrencySym`

given an asset policy we might want to know if it is the one being validated.

this is done by comparing it with the one specified in the datum field

```ts
const isOwnCurrSym = plet( account.currencySym.eqTerm );
```

##### `outIncludesNFT`

given a transaction output is useful to check if the value includes the NFT generated from the `accountFactory`;

this is done by checking that at least one of the attached value's entry satisfies `isOwnCurrencySym` for the policy.

```ts
 const outIncludesNFT = plet(
    pfn([ PTxOut.type ], bool )
    ( out => out.value.some( entry => isOwnCurrSym.$( entry.fst ) ) )
);
```

##### `ownOuts`

transaction outputs filtered by `isOwnOutput`;

##### `ownInputs`

transaction inputs filtered by `isOwnInput`;

##### updating the `Account` datum

> Between all the standard redeemers it must be checked that the datum fields `credentials`, `currencySym` and `state` remain unchanged compared to the current datum.

Reviewer 4: Not intended to do a semantic review (just fixing typos), but I could not help to observe:

If there is a requirement of X to be unchanged, I would strongly suggest to mention it explicitly everywhere needed.

We had this same "ensure that all non mentioned parts of a datum remain unchanged" in the [Hydra specification](https://hydra.family/head-protocol/core-concepts/specification/) and it resulted in at least one security vulnerability where forgot that check.

Author: That makes a lot of sense, will make sure to report it where needed

Author: went through the specification and this seems to be reported where needed, let me know if you spot any inconsistencies!

This might not be true for any implementation specific additional redeemer.

#### `Mint`

> the contract being called using the `Mint` redeemer MUST suceed only if the following conditions are met:

Reviewer 4: ```suggestion
the contract being called using the `Mint` redeemer MUST succeed only if the following conditions are met:
```

- the output with the sender's credentials must preserve the NFT

- the account fields `credentials`, `currencySym` and `state` are unchanged between sender input datum and sender output datum

- the account fields `credentials`, `currencySym` and `state` are unchanged between receiver input datum and receiver output datum

- `ownOuts` to be of lenght `2`

- every member of `ownOuts` must have at most 2 entries to prevent DoS by token spam.

- the input sender's amount field is greather or equal than the redeemer amount (`transfer.amount`)

- the output sender's amount field is equal to the input one minus `transfer.amount`

- the output receiver's amount field is equal to the input one plus `transfer.amount`

- the sender singned the transaction (included in `ctx.tx.signatories` if a `PPubKeyHash` or included a script input if a `PValidatorHash`)

any additional check can be made based on the `account.state` (implementation specific)

#### `Receive`

- `ownInputs` to be of lenght `2`

- the sender input (the `ownInput` with different `utxoRef` of the `validatingInput`)
being spent with `Transfer` redeemer

> - in the sender input `Transefer` redeemer the `to` field is equal to the `account.credentials` (where `account` is the receiver datum)

Reviewer 4: ```suggestion
- in the sender input `Transfer` redeemer the `to` field is equal to the `account.credentials` (where `account` is the receiver datum)
```

---

> CIP: _???

Reviewer: ```suggestion
CIP: ?
```
... for consistency with other drafts in the meantime

> Title: meta-assets (ERC20-like assets)

Reviewer: ```suggestion
Title: ERC20-like assets
```
... for consistency with PR title and other CIP titles.

Category: Tokens
Status: Proposed
Authors:
    - Michele Nuzzi <michele.nuzzi.204@gmail.com>
    - Harmonic Laboratories <harmoniclabs@protonmail.com>
    - Matteo Coppola <m.coppola.mazzetti@gmail.com>
Implementors: []
Discussions:

>     - https://github.com/cardano-foundation/cips/pulls/?

Reviewer: ```suggestion
    - https://github.com/cardano-foundation/CIPs/pull/444
```

Created: 2023-01-14
License: CC-BY-4.0
---

> <!-- Existing categories:
> 
> - Meta                   | For meta-CIPs which typically serves another category or group of categories.
> - Reward-Sharing Schemes | For CIPs discussing the reward & incentive mechanisms of the protocol.
> - Wallets                | For standardisation across wallets (hardware, full-node or light).
> - Tokens                 | About tokens (fungible or non-fungible) and minting policies in general.
> - Metadata               | For proposals around metadata (on-chain or off-chain).
> - Tools                  | A broad category for ecosystem tools not falling into any other category.
> - Plutus                 | Changes or additions to Plutus
> - Ledger                 | For proposals regarding the Cardano ledger
> - Catalyst               | For proposals affecting Project Catalyst / the Jörmungandr project
> 
> -->
> 
> # CIP-XXXX: meta-assets (ERC20-like assets)
> <!-- A reminder of the document's number & title, for readability. -->
> 

Reviewer: ```suggestion
```
H1 headers are generated from the YAML preamble and we don't want to fossilise this comment from the template which has been changing year by year.

## Abstract

> <!-- A short (\~200 word) description of the proposed solution and the technical issue being addressed. -->

Reviewer: ```suggestion
```

This CIP proposes a standard that if adopted would allow the same level of programmability of other ecosistems at the price of the token true ownership.

This is achieved imitating the way the account model works laveraging the UTxO structure adopted by Cardano.

## Motivation: why is this CIP necessary?

> <!-- A clear explanation that introduces the reason for a proposal, its use cases and stakeholders. If the CIP changes an established design then it must outline design issues that motivate a rework. For complex proposals, authors must write a Cardano Problem Statement (CPS) as defined in CIP-9999 and link to it as the `Motivation`. -->

Reviewer: ```suggestion
```
Sorry I should have noted how many of these template comments there were & simply told _you_ to get rid of them 😅

This CIP proposes a solution at the Cardano Problem Statement 3 ([CPS-0003](https://github.com/cardano-foundation/CIPs/pull/382/files?short_path=5a0dba0#diff-5a0dba075d998658d72169818d839f4c63cf105e4d6c3fe81e46b20d5fd3dc8f)).

If adopted it would allow to introduce the programmability over the transfer of tokens (meta-tokens) and their lifecycle.

The solution proposed includes (answering to the open questions of CPS-0003):

1 and 2) very much like account based models, wallets supporting this standard will require to know the address of the smart contract (validator)

3) the soultion can co-exsist with the exsiting native tokens

4) the implementaiton is possible without hardfork since Vasil

5) optimized implementations should not take significant computation, especially on transfers.

## Specification

> <!-- The technical specification should describe the proposed improvement in sufficient technical detail. In particular, it should provide enough information that an implementation can be performed solely on the basis of the design in the CIP. This is necessary to facilitate multiple, interoperable implementations. -->

Reviewer: ```suggestion
```
ditto

- the output receiver's amount field is equal to the input one plus `transfer.amount`

- the sender singned the transaction (included in `ctx.tx.signatories` if a `PPubKeyHash` or included a script input if a `PValidatorHash`)

any additional check can be made based on the `account.state` (implementation specific)

#### `Receive`

- `ownInputs` to be of lenght `2`

- the sender input (the `ownInput` with different `utxoRef` of the `validatingInput`)
being spent with `Transfer` redeemer

- in the sender input `Transefer` redeemer the `to` field is equal to the `account.credentials` (where `account` is the receiver datum)

> **NOTE**
>
> the sender input being an `ownInputs` element and being spent with `Transfer` redeemer 
> implies that the transfer calculation is running in that validation
>
> hence we don't need to perform the same calculation here

#### `ForwardCompatibility`

For the current version this redeemer SHOULD always fail.

The main purpose of the redeemer is to avoid breaking compatibilities for addtional implementation specific redeemers

## Rationale: how does this CIP achieve its goals?

> <!-- The rationale fleshes out the specification by describing what motivated the design and what led to particular design decisions. It should describe alternate designs considered and related work. The rationale should provide evidence of consensus within the community and discuss significant objections or concerns raised during the discussion.
> 
> It must also explain how the proposal affects the backward compatibility of existing solutions when applicable. If the proposal responds to a CPS, the 'Rationale' section should explain how it addresses the CPS, and answer any questions that the CPS poses for potential solutions.
> -->

Reviewer: ```suggestion
```
**No** you don't need the comment but **yes** please contribute a section as indicated in this comments.  You could paraphrase a bunch of reviewer comments & your own responses already in this PR and have a decent Rationale for this proposal... at least to start with.  In any case it _does_ need _something_.

Reviewer 2: Please do add applicable content here @michele-nuzzi, this is a requirement for us to be able to assign a number and then merge.

## Path to Active

### Acceptance Criteria
<!-- Describes what are the acceptance criteria whereby a proposal becomes 'Active' -->

> - having at least one instance of the smart contracts described on:
>     - mainnet
>     - preview testnet
>     - preprod testnet
> - having at least 2 different wallets integrating meta asset functionalities, mainly:
>     - displayning balance of a specified meta asset if the user provides the address of the respecive account manager contract
>     - transaction creation with `Transfer` redeemers

Reviewer: ```suggestion
- having at least one instance of the smart contracts described on:
    - mainnet
    - preview testnet
    - preprod testnet
- having at least 2 different wallets integrating meta asset functionalities, mainly:
    - displayning balance of a specified meta asset if the user provides the address of the respecive account manager contract
    - transaction creation with `Transfer` redeemers
```
Since these items will likely happen at different times, they should also have check-boxes like the _Implementation Plan_.  It will look more professional & quotable, and more consistent with other CIPs, if these have a capital letter at the beginning of each phrase, even though generally not complete sentences.

### Implementation Plan
<!-- A plan to meet those criteria. Or `N/A` if not applicable. -->

- [x] [PoC implementation](https://github.com/HarmonicLabs/erc20-like)
- [x] [showcase transactions](https://github.com/HarmonicLabs/erc20-like)
- [ ] wallet implementation 

## Copyright

> <!-- The CIP must be explicitly licensed under acceptable copyright terms. -->
> 
> [CC-BY-4.0]: https://creativecommons.org/licenses/by/4.0/legalcode
> [Apache-2.0]: http://www.apache.org/licenses/LICENSE-2.0

Reviewer: ```suggestion

This CIP is licensed under [CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/legalcode).
```
Recent edits to the CIP `.github` template make this a little easier to understand.

            style mntInStop  fill:#FFFFFF00, stroke:#FFFFFF00;
        end

        subgraph burned_asset
            direction LR
            brnInStart[ ] -.-x brnInStop[ ]
        
            style brnInStart fill:#FFFFFF00, stroke:#FFFFFF00;
            style brnInStop  fill:#FFFFFF00, stroke:#FFFFFF00;
        end

        style input/output fill:#FFFFFF00, stroke:#FFFFFF00;
        style ref_input fill:#FFFFFF00, stroke:#FFFFFF00;
        style minted_asset fill:#FFFFFF00, stroke:#FFFFFF00;
        style burned_asset fill:#FFFFFF00, stroke:#FFFFFF00;
    end
    style utxos fill:#FFFFFF00, stroke:#FFFFFF00;

    subgraph transactions
        direction LR

        subgraph Transaction
            .[ ]
            style . fill:#FFFFFF00, stroke:#FFFFFF00;
        end

        subgraph transaction
            direction LR
            start[ ] --> stop[ ]
        
            style start fill:#FFFFFF00, stroke:#FFFFFF00;
            style stop  fill:#FFFFFF00, stroke:#FFFFFF00;
        end
        style transaction fill:#FFFFFF00, stroke:#FFFFFF00;

    end
    style transactions fill:#FFFFFF00, stroke:#FFFFFF00;

    

```

### High level idea

> The core idea of the implementation is to emulate the ERC20 standard; where tokens are entries in a map with addresses (or credentials in our case) as key and integers (the balances) as value. ([see the OpenZeppelin implementation for reference](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/9b3710465583284b8c4c5d2245749246bb2e0094/contracts/token/ERC20/ERC20.sol#L16));

Reviewer 5: I'd like to suggest an alternative approach that leverages the [distributed DApp design](https://github.com/cardano-foundation/CIPs/pull/466). The idea is to give everyone their own DApp address - all addresses use the same plutus script for the payment credential and use a personal staking credential. Then, assets within a DApp address can only be sent to other addresses for that DApp.

Consider an "art vault" DApp that is meant to hold a user's art NFT collection. If Alice wants to send her SpaceBudz NFT (which is located inside her art vault) to Bob, she must send the NFT to Bob's art vault (the vault DApp address using Bob's staking credential). Trying to send the NFT to a non-vault address should fail.

This approach effectively creates a sub-ledger that can manage its own state in a distributed manor, and it enables adding configurable "controls" on how assets are traded between the DApp's addresses. For example, with the art vault DApp, every transfer can enforce a royalty payment; this would enable the cardano settlement layer to naturally enforce royalties. To do this, every transaction that spends an NFT from within the art vault "network" must reference the required input that contains the royalty information for that NFT (using something like [CIP-68](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0068)). The vault DApp would enforce the required royalties seen in the referenced input. I actually created a proof-of-concept of this a while ago, but had to put it aside due to time constraints (the documentation isn't complete).

Using the distributed approach, you can also freeze assets/addresses. The distributed DApp could require all transactions to reference a particular input (or set of inputs) and if it is missing or configured a certain way, the transaction is guaranteed to fail. These required reference inputs can be governed by some entity. So a stablecoin issuer can have a centrally controlled reference input that has a list of blacklisted addresses; any transaction that tries to spend from/to these addresses can be made to fail.

The benefits of the distributed approach are:
1. More scalable since you don't need to carry around global state with every transaction. All state is local.
2. The state is naturally enforced by the Cardano Blockchain - in other words, a transfer from Alice to Bob is already enforced/secured by Cardano. Instead, the distributed DApp just needs to ensure the assets can't inappropriately leave the DApp address network.
3. Other DApps can more easily "see" the current state for a user since the state is local. The local state makes it significantly easier for the distributed DApp to compose with other DApps. Compare this to smart contracts needing to look at the global state for everyone, including users not relevant to the transaction.

Below is are some images depicting what I mean:
```mermaid
---
title: Normal Operations (all txs reference GreenLight)
---
  flowchart TB
    GreenLight
    Alice --> Charlie
    Alice --> Mike
    Bob --> Charlie
    Bob --> Mike
    Mike --> Charlie
    Mike --> Alice
```
```mermaid
---
title: Emergency Stop (all txs are blocked due to 'GreenLight' now being 'RedLight')
---
  flowchart TB
    RedLight
    Alice --X--> Charlie
    Alice --X--> Mike
    Bob --X--> Charlie
    Bob --X--> Mike
    Mike --X--> Charlie
    Mike --X--> Alice
```
```mermaid
---
title: Freeze Mike (all txs are allowed except for those to/from Mike)
---
  flowchart TB
    GreenExceptMike
    Alice --> Charlie
    Alice --X--> Mike
    Bob --> Charlie
    Bob --X--> Mike
    Mike --X--> Charlie
    Mike --X--> Alice
```

Reviewer 6: Haha seeing this knowing I worked on similar design contract is kinda funny. My idea was more related to account abstraction specifically, but it shares a bunch of similarities.

Author: @fallen-icarus I'm afraid I need to ask you to explain in simpler terms.

My current understanding of what you are proposing is that the state is somewhat centralized, which would scale linearly with the number of users.

The current CIP113 proposed standard only needs the sate of the parties involved, so scaling is not an issue.

Reviewer 5: There is no centralized state whatsoever. There is only (possibly) centralized control of the configurations.

There are two main differences between what I am suggesting and this CIP:
1. I am suggesting to store the UTxOs in **different** addresses; not a single manager address. Instead, each address would use the same payment credential (the smart contract required to control spending of the UTxOs), and unique staking credentials.
2. The tokens are **actual tokens**, not token-like.

Since the tokens are actual native assets, there is no need to re-invent the transfer logic; you can just leverage Cardano's transfer mechanisms. The spending of these tokens can be further constrained by configurable reference inputs.

Author: Yes both your points is what the current standard proposes.

the `transferManager` is the same payment credentials, the user uses the stake credentials to sign

tokens are actual CNTs in the contract.

The part of the code you cited is a left over from previous the specs

Reviewer 5: > the transferManager is the same payment credentials, the user uses the stake credentials to sign

Ah, okay. I see that now. Thanks for clarifying. 

What is the point of the `stateManager` contract in the [transfer diagram](https://github.com/cardano-foundation/CIPs/blob/f13516f144df991de14eca6dde9c4319bd98c791/CIP-meta-assets%20(ERC20-like%20assets)/README.md#transfer)? What "account" states are being updated? The tokens are actually transferred so why isn't this enough to reflect the new ownerships?

Author: We want to enforce logic for the entire account rather than per-utxo

Imagine we want to freeze an account

without a `stateManager` we can still send utxos to that credential that have different state.

Instead with a "global" (not really but pass me the term) state for the account we make the change once and the possibly many utxos of the credentials are all freezed

Reviewer 5: I misread the diagram... I thought the `stateManager` inputs were normal inputs, but I see now they are reference inputs. The state manager actually isn't doing anything during the transfer.

> Yes both your points is what the current standard proposes.

I see it now. Thanks for answering my questions!

Unlike the ERC20 standard; this CIP:

- allows for multiple entries with the same key (same credentials can be used for multiple accounts, though not particularly useful)
- DOES NOT include an equivalent of the `transferFrom` method; if needed it can be added by a specific implementation but it won't be considered part of the standard.
- allows for multiple inputs (from the same sender, scripts and multisigs MAY also be used) and multiple outputs (possibly many distinct reveivers)

> **NOTE: Multiple Inputs**
>
> the UTxO model allows for multiple transfers in one transaction
>
> this would allow for a more powerful model than the account based equivalent but implies higher execution costs
>

> > with the goal of keeping the standard simple we allow only a single sender

Author: Yes, only a single sender allowed so far, but open to discussions

>
> we MUST however account for multiple inputs from the same sender due to
> the creation of new UTxOs in receiving transactions.

### Design

with the introduction of [CIP-69](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0069) in Plutus V3 the number of contracts required are only 2, covering different purposes.

1) a `stateManager` contract
    - with minting policy used to proof the validity of account's state utxos
    - spending validator defining the rules to update the states
2) a `transferManager` parametrized with the `stateManager` hash

>     - having minting policy for the tokens to be locked in the spending validator

Reviewer 5: Can this be made more general? Having the `transferManager` also mint the tokens means users will have a separate address for every programmable token. This can quickly become a nightmare for frontends to integrate. It would be better if there was a single `transferManager` that could be used with all programmable tokens.

Programmable tokens don't necessarily need to be minted by the `transferManager`. Instead, they can be deposited into the `transferManager` by the entity minting the token. For example, a user selling NFTs can ensure that all NFTs sold go to a `transferManager` address.

There may also be use cases where a token should be allowed to leave the `transferManager`. For example, lets say there is an art NFT that has been collecting royalties for a while and now the royalty owners would like to release the NFT from the royalty requirement (there may be legal/tax benefits to this in the future). This approach doesn't seem to have a "release" mechanism.

A "release" mechanism could also be useful for upgrading the `transferManager`. If a new version comes out, users can possibly transfer their assets from `transferManager` v1 to `transferManager` v2 (depending on the control utxos). The current design seems to prevent upgrades like this. If this CIP actually does support upgrades, it definitely needs to mention how upgrades are possible. Upgrades don't seem to be mentioned anywhere.

    - spending validator for the validation of single utxos (often just forwarding to the withdraw 0)

>     - certificate validator to allow registering the stake credentials (always succeed MAY be used) and specify the rules for de-registration (always fail MAY be used, but some logic based on the user state is RECOMMENDED)

Author: Yes you are right, de-registering should not be based on the user state

>
>

```ts
const TransferOutput = pstruct({
    TransferOutput: {
        credential: PCredential.type,
        amount: int,
        stateIndex: int
    }
});

const TransferRedeemer = pstruct({
    Transfer: {
        totInputAmt: int,
        outputs: list( TransferOutput.type ),
        stateIndex: int
        firstOutputIndex: int,
        // inputIndicies: list( int ), // filter
    }
});
```

#### `SingleTransferRedeemer`

redeemer to be used on a single utxo of the `transferManager`;

```ts
const SingleTransferRedeemer = pstruct({
    Transfer: {},
    Burn: {},
});
```

### Transactions

#### New Account Generation Transaction

```mermaid
flowchart LR

    stateManagerPolicy[(state manager policy)]

    subgraph transaction
        .[ ]
        style . fill:#FFFFFF00, stroke:#FFFFFF00;
    end

    stateManagerContract[state manager]

    stateManagerPolicy -. validity NFTs .-> transaction 
    --o stateManagerContract
```

The purpose of the "New Account generation" transaction is to create a new account for a user and validate the initial state of an account.

The validation is performed in the minting policy.

> The minting policy MUST succeed if only one asset is created under the policy respecting the naming convention explained below, and MAY succeed if other assets are created as long as their name is NOT of length 32.

Author: to avoid potential conflict with the resulting hash

very remote but better avoid

should not be a very useful feature anyway

The transaction MUST have an output going to the `transferManager` contract with the correct [`TransferManagerDatum`](#TransferManagerDatum) **inline** datum.
 
The first field of the datum MUST be the name of the NFT present on the reference input coming from the state manager.

Additional, implementation-specific, arbitrary logic (eg. capped supply, or one-time mints, etc. ) MAY be added.

#### Burning Tokens

```mermaid
flowchart LR

    subgraph transaction
        .[ ]
        style . fill:#FFFFFF00, stroke:#FFFFFF00;
    end

    stateManager[state manager]
    transferManagerContract[transfer manager]

    stateManager -. sender account state .-o transaction

    transferManagerContract -- Burn Redeemer --o  transaction -..-x a[ ]

    style a fill:#FFFFFF00, stroke:#FFFFFF00;
```

If an utxo in the `transferManager` is spent with `Burn` redeemer, the **entire** amount of the programmable token on that utxo MUST be burnt.

Only **one (1)** utxo from the `transferManager` is allowed to be spent in a burn transaction.

> If a user desires to burn a different amount than the one currently present they should break the operation in 2 different transactions:

Reviewer 5: Why only one utxo per transaction? This seems unnecessarily restrictive. Why can't multiple utxos be spent? The entire balance from each utxo spent can be burned. The observer can tell which utxos are being used for burning based on the redeemers used to spend them.

```mermaid
flowchart LR
    transferManagerObserver([transfer manager observer])
    stateManagerContract[state manager]
    transferManagerContract[transfer manager]

    sender((sender))

    A[transfer manager]
    B[transfer manager]
    same[transfer manager]

    subgraph transaction
        direction LR
        .[ ]
        _[ ]
        z[ ]
        style . fill:#FFFFFF00, stroke:#FFFFFF00;
        style _ fill:#FFFFFF00, stroke:#FFFFFF00;
        style z fill:#FFFFFF00, stroke:#FFFFFF00;
    end

    transferManagerObserver -. validates inputs .-> transaction

    stateManagerContract -. sender account state .-o transaction
    stateManagerContract -. receiver A account state .-o transaction
    stateManagerContract -. receiver B account state .-o transaction

    transferManagerContract --o transaction
    transferManagerContract -- possibly --o transaction
    transferManagerContract -- many --o transaction
    transferManagerContract -- inputs --o transaction
    transferManagerContract --o transaction

    sender -. signs .-> transaction

    transaction -- stake creds A --o A
    transaction -- stake creds B --o B
    transaction -- change (if needed) --o same
```

> The "Transfer" transaction allows the transfer of programmable tokens from one account to another within the `transferManager` contract. This transaction involves the following components:

Reviewer 2: ```suggestion
The `Transfer` transaction allows the transfer of programmable tokens from one account to another within the `transferManager` contract. This transaction involves the following components:
```

The transaction MUST include:

1. One or more inputs from the sender's account in the `transferManager` with `SingleTransferRedeemer` for each UTxO spent.
2. Outputs to the receives' accounts in the `transferManager` as specified in the `TransferRedeemer` of the transfer manager observer.
3. The `TransferRedeemer` for the `transferManager` (as observer) contract.
4. one `Account` reference input for each of the parties involved (sender + receivers).

The information in the `TransferRedeemer` MUST be validated by the observer.

That includes that the sum of the programmable tokens coming from the `transferManager` is stirctly equal to `totInputAmt`.

There is one output for each element of the 2nd field (`outputs`) starting at the index specified by the 4th field (`firstOutputIndex`), in the same order.

Each of the outputs going to the `transferManager` MUST have `TransferManagerDatum` **inline** datum, and MUST make sure the first field of the datum matches the token name of the NFT present on the respective reference input `Account`.

## Rationale: how does this CIP achieve its goals?

> <!-- The rationale fleshes out the specification by describing what motivated the design and what led to particular design decisions. It SHOULD describe alternate designs considered and related work. The rationale SHOULD provide evidence of consensus within the community and discuss significant objections or concerns raised during the discussion.
> 
> It MUST also explain how the proposal affects the backward compatibility of existing solutions when applicable. If the proposal responds to a CPS, the 'Rationale' section SHOULD explain how it addresses the CPS, and answer any questions that the CPS poses for potential solutions.
> -->

Reviewer: ```suggestion
```
(artefact from template comment scaffold)

The [first proposed implementation](https://github.com/cardano-foundation/CIPs/pull/444/commits/525ce39a89bde1ddb62e126e347828e3bf0feb58) (which we could informally refer as v0) was quite different by the one shown in this document

Main differences were in the proposed:
- [use of sorted merkle trees to prove uniqueness](https://github.com/cardano-foundation/CIPs/pull/444/commits/525ce39a89bde1ddb62e126e347828e3bf0feb58#diff-370b6563a47be474523d4f4dbfdf120c567c3c0135752afb61dc16c9a2de8d74R72) of an account during creation;

> - account credentials as asset name

Author: using the credentials for the name could not guarantee the unique association of an account to a set of utxos

a user could create as many accounts with the same name, with different states (or control  data, or config, whatever you like) which could represent a security vulnerability from the perspective of the spending logic

> - account credentials as asset name

Reviewer 5: Currently, the `transferManager` is parameterized by the `stateManager`. I haven't fully thought this through, but what if it wasn't and instead the `stateManager` hash was in the `transferManager` datum? It could be something like:
1. `accountManager` - a (new) minting policy where the token name is the user's credential. This enables the DApp entity to see all registered credentials.  This `accountManager` does *not* control any state; it just enables using the list of associated tokens as a list of registered users.
2. `stateManager` - the current `stateManager` except it is parameterized by the `accountManager` hash *and* the user's credential. So all users get their own `stateManager` minting policy for the programmable token. The DApp entity can see the credential in the registered list (tokens for `accountManager`) and can then lookup all associated `stateManager` tokens for that credential. Minting new `state` tokens requires referencing the `account` token for that credential.
3. `transferManager` - the current `transferManager` except it is not parameterized by anything. Instead, the `stateManager` hash is also included in the `transferManager` datum. The `accountManager` hash could also go in the datum. This contract does **not** mint the programmable tokens. Instead, they programmable tokens will be deposited into it.

This approach could make the `transferManager` general purpose so each user only needs one address for all programmable tokens. This approach also makes it easy for users to check their statuses since *all* tokens minted with a given `stateManager` (parameterized with their credential) belongs to them. In other words, they can find all of their state UTxOs using just their credential. Thoughts?

> - account credentials as asset name

Author: > This approach could make the transferManager general purpose so each user only needs one address for all programmable tokens

I find it hard to believe that many tokens will adhere to the same spending rules

> - account credentials as asset name

Reviewer 5: > I find it hard to believe that many tokens will adhere to the same spending rules

I don't think they will, but the differences could possibly be encoded into the state UTxOs. Currently, the lack of generality is a deal breaker IMO. Do you expect users to have a separate address for every stablecoin, every NFT collection, every goverment issued DID, etc? Do you expect frontends to actually add support for possibly an infinite number of addresses? I don't. You might get a few addresses to be supported, but that means the frontends effectively act as gatekeepers for which tokens can be programmable since unsupported tokens effectively don't exist. Perhaps there can't be a universal `transferManager`, but perhaps 2 or 3 different `transferManagers` can cover most use cases.

After a first round of community feedback, a [reviewed stadard was proposed](https://github.com/cardano-foundation/CIPs/pull/444/commits/f45867d6651f94ba53503833098d550326913a0f) (which we could informally refer to as v1).
[This first revision even had a PoC implementation](https://github.com/HarmonicLabs/erc20-like/commit/0730362175a27cee7cec18386f1c368d8c29fbb8), but after further feedback from the community it was noted that the need to spend an utxo on the receiving side could cause UTxO contention in the moment two or more parties would have wanted to send a programmable token to the same receiver at the same time.

The specification proposed in this file addresses all the previous concerns.

The proposal does not affect backward compatibilty being the first proposing a standard for programmability over transfers;

exsisting native tokens are not conflicting for the standard, instead, native tokes are used in this specification for various purposes.

## Path to Active

### Acceptance Criteria

> <!-- Describes what are the acceptance criteria whereby a proposal becomes 'Active' -->

Reviewer: ```suggestion
```
(artefact from template comment scaffold)

- having at least one instance of the smart contracts described on:
    - mainnet
    - preview testnet
    - preprod testnet
- having at least 2 different wallets integrating meta asset functionalities, mainly:
    - displayning balance of a specified meta asset if the user provides the address of the respecive account manager contract
    - independent transaction creation with `Transfer` redeemers

### Implementation Plan

> <!-- A plan to meet those criteria. Or `N/A` if NOT applicable. -->

Reviewer: ```suggestion
```

Category: Tokens
Status: Proposed
Authors:
    - Michele Nuzzi <michele.nuzzi.2014@gmail.com>
    - Matteo Coppola <m.coppola.mazzetti@gmail.com>
    - Philip DiSarro <philipdisarro@gmail.com>
Implementors: []
Discussions:

>     - https://github.com/cardano-foundation/CIPs/pull/444

Reviewer: ```suggestion
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/444
    - https://github.com/cardano-foundation/CIPs/pull/944
```
